dolosse/generic_json_consumer.py

```diff
@@ -0,0 +1,142 @@
+"""Reads json data from a kafka topic and graphs it"""
+import json
+import time
+import sys
+import curses
+from matplotlib import pyplot
+import kafka
+
+
+def populate(variables, decode,
+             x_axis_data, prepend=""):
+    """Takes the data from the dict and populates the variables"""
+    for other_key in decode:
+        valid = True
+        name = prepend+other_key
+        try:
+            following = float(decode[other_key])
+        except (ValueError, TypeError):
+            valid = False
+        if isinstance(decode[other_key], dict):
+            populate(variables,
+                     decode[other_key],
+                     x_axis_data,
+                     name+": ")
+        else:
+            if name in variables:
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(
+                        variables[name][-1])
+            else:
+                variables[name] = [0]*(len(x_axis_data)-1)
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(0)
+                    #Remember, the list might still be empty here
+
+
+
+
+
+def check():
+    """Checks that the right number of args are present"""
+    topic = "default"
+    if (len(sys.argv)) < 2:
+        print("Please provide a topic name on the command line!")
+        print("Using default topic name 'default'...")
+    else:
+        topic = sys.argv[1]
+    return topic
+
+
+def redraw(figures, key):
+    """Redraw the given figure from a dict of figures"""
+    figures[key].canvas.draw()
+    figures[key].canvas.flush_events()
+
+
+
+
+
+def main():
+    """Main function - reads json data from a kafka topic and graphs it"""
+
+    screen = curses.initscr()
+    curses.cbreak()
+    screen.nodelay(True)
+
+    topic = check()
+
+    data = {}
+    line = {}
+    plots = {}
+    figures = {}
+    axes = {}
+    x_axis_data = []
+    count = 0
```

Comment:
Instantiate these at first use unless they're first used in a smaller scope (e.g. inside a for loop)

Reply:
They are first used inside the "for message in consumer" loop; they need to have the relevant types (mostly dicts, one list, on integer equalling zero) before entering the loop, because their first use assumes that they have those classes. For example, 'data' is first used in a call to 'data.update()' - a call that will only have the expected effect if 'data' is already a dict.

If I don't instantiate them before the loop, then I'm not quite sure where to do it that makes sense.

Reply:
Have since managed to find a way to live without the list as an explicit variable.